source/agora/utils/SCPPrettyPrinter.d

```diff
@@ -0,0 +1,606 @@
+/*******************************************************************************
+
+    Contains pretty-printing routines specialized for SCP types.
+
+    As the integration test-suite imports into PrettyPrinter,
+    we want to avoid adding a dependency to SCP.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.utils.SCPPrettyPrinter;
+
+import agora.common.Types;
+import agora.common.Amount;
+import agora.common.Config;
+import agora.common.Serializer;
+import agora.common.Types;
+import agora.common.crypto.Key;
+import agora.consensus.data.Block;
+import agora.consensus.data.ConsensusData;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import agora.utils.PrettyPrinter;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types : StellarHash = Hash;
+alias StellarValue = Value;
+
+import std.algorithm;
+import std.format;
+import std.range;
+
+/// Returns:
+/// A formatting struct for an SCPEnvelope
+public auto scpPrettify (in SCPEnvelope env, in GetQSetDg get_qset = null) nothrow
+{
+    return SCPEnvelopeFmt(env, get_qset);
+}
+
+/// Formatting struct for `SCPBallot`, deserializes Value types as ConsensusData
+public struct SCPBallotFmt
+{
+    private const(SCPBallot) ballot;
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @trusted nothrow
+    {
+        try
+        {
+            formattedWrite(sink,
+                "{ counter: %s, ",
+                this.ballot.counter);
+
+            try
+            {
+                formattedWrite(sink,
+                    "value: %s }",
+                    prettify(  // cast: deserializer should take const(ubyte)[]
+                    (cast(ubyte[])this.ballot.value[]).deserializeFull!ConsensusData));
+            }
+            catch (Exception ex)
+            {
+                formattedWrite(sink, "value: <un-deserializable> }");
+            }
+        }
+        catch (Exception ex)
+        {
+            assert(0, ex.msg);
+        }
+    }
+}
+
+/// SCP Quorum set getter delegate
+private alias GetQSetDg = SCPQuorumSetPtr delegate (
+    ref const(StellarHash) qSetHash);
+
+/// Formatting struct for a quorum StellarHash => QuorumConfig through the use
+/// of a quorum getter delegate
+private struct QuorumFmt
+{
+    private const(StellarHash) hash;
+    private const(GetQSetDg) getQSet;
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink)
+        @trusted nothrow
+    {
+        try
+        {
+            SCPQuorumSetPtr qset;
+            if (this.getQSet !is null)
+                qset = this.getQSet(this.hash);
+
+            if (qset !is null)
+            {
+                auto qconf = toQuorumConfig(*qset.ptr);
```

Comment:
Isn't that a `null` dereference ? `toQuorumConfig` takes a `ref`

Reply:
One was using `alias this` (`shared_ptr`) and the other explicit dereference. I'll simplify it.